src/lib/rengo_balancer.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2012-2022  Online-Go.com
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { errorAlerter } from "misc";
+import * as player_cache from "player_cache";
+import { put } from "requests";
+
+type Participant = {
+    user_id: number;
+    rating: number;
+};
+
+// Convert player IDs to Participants with ratings.
+async function toParticipants(playerIDs: number[]): Promise<Participant[]> {
+    const participants = [];
+    const required_fields = ["rating"];
+    for (const id of playerIDs) {
+        // TODO is there a way to fetch multiple players at once?
+        const entry = await player_cache.fetch(id, required_fields);
```

Comment:
Oh I see what the comment was about - `await` may cause batching not to happen.  In order to parallelize the async calls, it is good practice to collect a list of promises, and then combine them with `Promise.all()`. Something like:

```
const promises = [];
for (const id of ids) {
    const participant_promise = fetch(id).then((entry) => {
        return {user_id: entry.id, rating: entry.rating};
    };
    promises.push(participant_promise);
}
return Promise.all(promises);
```

(irrelevant parts omitted for brevity)

Reply:
Thanks, I've added these changes.